ksqldb-engine/src/main/java/io/confluent/ksql/engine/EngineContext.java

```diff
@@ -238,10 +335,17 @@ void registerQuery(final QueryMetadata query) {
       }
 
       persistentQueries.put(queryId, persistentQuery);
-      metaStore.updateForPersistentQuery(
-          queryId.toString(),
-          persistentQuery.getSourceNames(),
-          ImmutableSet.of(persistentQuery.getSinkName()));
+      if (createAsQuery) {
+        createAsQueries.put(persistentQuery.getSinkName(), queryId);
+      } else {
+        final Iterable<SourceName> allSourceNames = Iterables.concat(
+            Collections.singleton(persistentQuery.getSinkName()),
+            persistentQuery.getSourceNames()
+        );
+
+        allSourceNames.forEach(sourceName ->
+            otherQueries.computeIfAbsent(sourceName, x -> new HashSet<>()).add(queryId));
```

Comment:
I noticed that `otherQueries` is thread safe, but now we're putting non-threadsafe entries inside it (`HashSet`). If it needs to be thread safe, we should put in a concurrent data structure (I'm guessing just `Collections.synchronized(new HashSet<>())` is good enough)

Reply:
Done